### ISSUE TITLE
Implement CLI command for API Gateway management

### ISSUE THREAD

**Issue by csantanapr** [apache/incubator-openwhisk#1495]
The CLI will provide to the user an ability to configure APIs using the WSK CLI.

For MVP we are looking at the following commands specification as a basic starting point:

### wsk api -h
```
wsk api -h
work with APIs
Usage:
  wsk api [command]

Available Commands:
  create      create a new API
  get         get API details
  delete      delete an API
  list        list APIs
```

### wsk api create –h
```
wsk api create –h
create a new API
Usage:  
   wsk api create [BASE_PATH] API_PATH API_VERB ACTION [flags]

Flags: 
  -n, --apiname string           Friendly name of the API; ignored when CFG_FILE is specified (default BASE_PATH)  
  -c, --config-file CFG_FILE     CFG_FILE containing API configuration in swagger JSON format

Examples:
$ wsk api create –n “My Pet Store” /store /pets get    getPet
$ wsk api create –n “My Pet Store” /store /pets post   createPet 
$ wsk api create /store /pets                   put    updatePet 
$ wsk api create /store/pets                    delete deletePet
$ wsk api create –c petstore-swagger.json 
```

### wsk api delete –h
```
wsk api delete –h
delete an API
Usage:  
  wsk api delete BASE_PATH | API_NAME [API_PATH [API_VERB]] [flags]

Examples:
$ wsk api delete “My Pet Store”
$ wsk api delete /store
$ wsk api delete “My Pet Store” /pets
$ wsk api delete “My Pet Store” /pets put
$ wsk api delete /store /pets get
```

### wsk api list –h
```
wsk api list –h
list APIs
Usage:  
  wsk api list [[BASE_PATH | API_NAME] [API_PATH [API_VERB]] [flags]

Flags:  
-l, --limit LIMIT   only return LIMIT number of actions from the collection (default 30)  
-s, --skip SKIP     exclude the first SKIP number of actions from the result

Examples:
$ wsk api list
Action               Verb          API Name      URL
getPet               get           My Pet Store  https://699da.api-gw.mybluemix.net/store/pets
createPet            post          My Pet Store  https://699da.api-gw.mybluemix.net/store/pets
depositCheck         post          My Bank       https://699da.api-gw.mybluemix.net/bank/checks

$ wsk api list “My Pet Store”
Action               Verb          API Name      URL
getPet               get           My Pet Store  https://699da.api-gw.mybluemix.net/store/pets
createPet            post          My Pet Store  https://699da.api-gw.mybluemix.net/store/pets

$ wsk api list /store /pets
Action               Verb          API Name      URL
getPet               get           My Pet Store  https://699da.api-gw.mybluemix.net/store/pets
createPet            post          My Pet Store  https://699da.api-gw.mybluemix.net/store/pets
```

### wsk api get –h
```
wsk api get –h
get API details

Usage:  
  wsk api get BASE_PATH | API_NAME [flags]

Flags:  -f, --full   display full API configuration details

$ wsk api get “My Pet Store” > petstore-swagger.json
$ wsk api get /store > petstore-swagger.json
$ wsk api get “My Pet Store” –-full
{
"namespace": "guest",
"gwApiUrl": "https://699da.api-gw.mybluemix.net/store",
"tenantId": "699dac50-8601-4c6b-901e-28f721ef0351",
"apidoc": {
   "swagger": "2.0",
   "basePath": "/store",
    "info": {"title": ”My Pet Store","version": "1.0.0"},
   "paths": {
     "/pets": {
       “get”: {
         "x-ibm-op-ext": {
           "actionName": ”getPet", 
           "actionNamespace": "whisk.system",
           "backendMethod": "POST",
           "backendUrl": "https://openwhisk.ng.bluemix.net/api/v1/namespaces/whisk.system/actions/getPet",                        "policies": [
           ”policies”:[
```

### demo wsk api
Here is a quick demo on how to use the WSK CLI to configure your API GW

Lets create an API for a Bank name `Santana Bank` on basePath `/bank` to manage two type of resources `checks` and `accounts`
```
$ wsk api create -n "Santana Bank" /bank /checks get /whisk.system/utils/echo
ok: created api /checks GET for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
$ wsk api create -n "Santana Bank" /bank /checks post /whisk.system/utils/echo
ok: created api /checks POST for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
$ wsk api create -n "Santana Bank" /bank /checks put /whisk.system/utils/echo
ok: created api /checks PUT for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
$ wsk api create -n "Santana Bank" /bank /checks delete /whisk.system/utils/echo
ok: created api /checks DELETE for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
$ wsk api create -n "Santana Bank" /bank /accounts post /whisk.system/utils/echo
ok: created api /accounts POST for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
$ wsk api create -n "Santana Bank" /bank /accounts delete /whisk.system/utils/echo
ok: created api /accounts DELETE for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
$ wsk api create -n "Santana Bank" /bank /accounts get /whisk.system/utils/echo
ok: created api /accounts GET for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
$ wsk api create -n "Santana Bank" /bank /accounts put /whisk.system/utils/echo
ok: created api /accounts PUT for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
```
Let's list app APIs in our namespace

```
$ wsk api list
ok: apis
Action                           Verb             API Name  URL
whisk.system/utils/echo          post         Santana School  http://21ef035.api-gw.example.com/school/students
whisk.system/utils/echo           put         Santana School  http://21ef035.api-gw.example.com/school/students
whisk.system/utils/echo        delete         Santana School  http://21ef035.api-gw.example.com/school/students
whisk.system/utils/echo           get         Santana School  http://21ef035.api-gw.example.com/school/students

whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
```

Let's list only the ones for basePath `/bank`
```
$ wsk api list /bank
ok: apis
Action                           Verb             API Name  URL
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
```

We can also use API Name `Santana Bank` to list 
```
$ wsk api list "Santana Bank"
ok: apis
Action                           Verb             API Name  URL
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
```

Let's list only the ones for basePath `/bank` and API Path `/checks`
```
$ wsk api list /bank /checks
ok: apis
Action                           Verb             API Name  URL
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
```

Let's export API named `Santana Bank` into a swagger file that we can customize our APIs and store in Source Control to be use in continues integration and deployment (CI/CD) 
```
$ wsk api get "Santana Bank" > bank-swagger.json
$ vi bank-swagger.json
```
Let's test the swagger file by deleting the API name `Santana Bank`
```
$ wsk api delete "Santana Bank"
ok: deleted api Santana Bank
$ wsk api list
ok: apis
Action                           Verb             API Name  URL
```

Now let's restore the API named `Santana Bank` by using the file `bank-swagger.json`
```
$ wsk api create --config-file bank-swagger.json
ok: created api /accounts delete for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
ok: created api /accounts get for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
ok: created api /accounts post for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
ok: created api /accounts put for action utils/echo
http://21ef035.api-gw.example.com/bank/accounts
ok: created api /checks get for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
ok: created api /checks post for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
ok: created api /checks put for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
ok: created api /checks delete for action utils/echo
http://21ef035.api-gw.example.com/bank/checks
```

We can verify that the API is recreated
```
$ wsk api list
ok: apis
Action                           Verb             API Name  URL
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/accounts
whisk.system/utils/echo        delete         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo           get         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo          post         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
whisk.system/utils/echo           put         Santana Bank  http://21ef035.api-gw.example.com/bank/checks
```

Now just for fun lets deposit a check for `$1,000,000.99` with a HTTP `POST` using the API Gateway
```
$  curl -X POST -d '{"amount":"1000000.99", "account":"Carlos"}' http://21ef035.api-gw.example.com/bank/checks
{
  "amount": "1000000.99",
  "account": "Carlos"
}
```

**Comment by csantanapr** [comment]
CC @mbehrendt here is the issue you requested

**Comment by mbehrendt** [comment]
thx @csantanapr .

i'm very interested in feedback from the community on this -- looking forward to getting all your thoughts.

**Comment by jthomas** [comment]
This is going to be a great feature, looking forward to trying it out. 

Two things I noticed.... 

- Does the API endpoint need both a base path and api path? Is it possible to just have my API GW URLs at /action? 

- It is possible to have wildcard routes in the URLs? I'm thinking how to handle /app/users/1 and /app/users/2.

- It would be nice if the CLI helped you define multiple HTTP methods without requiring each as a separate command. 

```
wsk api create /store/pets  * my_action 
```

Also, I find the CLI output data a bit difficult to quickly scan. If I want to see the paths and associated actions from the CLI I have to read all the way to end of each URL field. The URL field has a lot of extra information that I have to parse to understand this. Could we add a 'path' field? I would like to see "path, verb, action, URL" in the output table.

**Comment by mdeuser** [comment]
@jthomas - Thanks for the comments and suggestions! 

- Currently, when creating an API endpoint mapping, the basepath argument is optional and will default to '/'; however, a relative path is required.  If I understand your example correctly, a single argument of `/action` would be the equivalent of a basepath of `/`, an api path of `/action` mapping to an action called `action`.  Fully qualified action names would result in api paths like `/whisk.system/utils/echo`.  Is my understanding correct?
- Re wildcard routes.  Maybe what you are asking for is how to create an api path that takes a path parameter?  At the moment that is not supported by the API GW; but we are aware of the need for this feature.
- Not currently.  At the moment, there's no way for an action to discern which HTTP method was used to invoke the action.  We have some thoughts on how to surface this information to the action; at that time, I think your suggestion would be a nice enhancement.
- Regarding the `wsk api list` output.  Would you like to see this additional path include both the base path and the api path, or just the api path?

**Comment by jthomas** [comment]
- Thanks for explaining the background to the base and api paths. 

  It seems like the base and api path stuff is an implementation detail. I think it would improve the developer experience if you just let the developer provide a single URL path and have the tooling handle splitting it into base and api paths. 

  It looks like the tool expects to support this format already.
  I would remove the other styles (separate base and api paths) to reduce confusion.

```
wsk api create /store/pets                    delete deletePet
```

- Regarding HTTP methods. Can someone define more than one HTTP method for a URL (even though the action can't tell the method)?

- Regarding `wsk api list` I'd suggest showing the full path, combining the base and api path as per my recommendation above.

**Comment by mdeuser** [comment]
- I think the challenge is to determine how much API-centric details to surface while configuring the action APIs.

 Due the the many possible API GW configuration properties, The CLI will probably not be able to provide args and options to fully configure every API option - in these cases, the API configuration can be exported (via the wsk CLI) in OpenAPI (i.e. Swagger) format, edited and then re-imported (via the wsk CLI).  The OpenAPI file will have basepath and path fields; so some understanding of this decomposition may be needed.

 Additionally, some folks may want more granular control of those values, so the CLI exposes them.  One possible way to support API-centric and OpenWhisk-centric developers is for the `wsk api create` command to accept a single argument in the form `/ACTION` - in addition to the allowing explicit basepath, api path and action arguments.  Would need to think a bit on how this would trickle out to the other commands.

- Right now, separate CLI commands must be used to assign separate HTTP methods to the same action.

- Re `wsk api list` suggestion.  I've opened issue #1534.

**Comment by csantanapr** [comment]
Thanks for the feedback @jthomas 
- splitting into `basePath` and `path` 
About the UX for getting started, the user can provide a single string like `/cats/dogs/rain` but to have a solution that we can configure the apigateway service and be able to generate a swagger file/spec. we have to split it into a `basePath` and `path` these two fields are entity required, we can only assume how to split `/cats/dogs/rain`  currently we are splitting into `/` and `/cats/dogs/rain` this is if the user doesn't go explicit, thinking that would be the more conservative way. 
So if user creates a `get` on `/cats/dogs/rain` and we decide to split into `basePath=/cats/dogs` and `path=/rain`. 
If the user creates another api operation `post` with `/cats/dogs/rain/monkeys` it will create a second swagger spec and API definition with `basePath=/cats/dogs/rain` and `path=/monkeys`
It might be an odd situation, but it might be OK since they always have the option to be explicit.
With using `/` then it will create a single API swagger and every api operation command were are not explicit will be under this API swagger.
The other way around every top parent path will create a new API swagger as long the first portion is the same.
I think is something to think about what would be the best default.

- Using `*` for method as an override
I think is a cool idea very nice shortcut, take into account that all verbs (options, head, get, put, post, delete, etc..) supported will be created, but I think for getting started it might be a convenience and also demo wise a good thing, before starting to talk the other more advance customizations and control like the swagger and being explicit.

**Comment by csantanapr** [comment]
@jthomas @mdeuser scratch my assumption 
~~So if user creates a get on /cats/dogs/rain and we decide to split into basePath=/cats/dogs and path=/rain.~~

basPath can't have subpaths, so allowing the short version to specify a Path I think is a good idea like
`wsk api create /store/pets get petaction`
It will be parsed basePath=`/store` and path=`/pets`

**Comment by mbehrendt** [comment]
@csantanapr @mdeuser @jthomas wouldn't the current approach of making the base path an optional argument satisfy james requirement (which i agree is a very valid one) ?

**Comment by mdeuser** [comment]
CLI path specification discussion has moved to #1535

**Comment by rabbah** [comment]
Suggest making the verb optional with get as default. --method to override. I'd expect this is the common case and cuts down on number of positional args which can be confusing to someone just starting to use this or kick the tires.

**Comment by mdeuser** [comment]
@csantanapr - ok to close this issue now that the `wsk api` command is available?

**Comment by mdeuser** [comment]
@csantanapr - ok to close?
